### PR TITLE
LVAE losses tests

### PR DESCRIPTION
**Pull request by CatEek**
### Description

- **What**:  Moved losses/NMs/LL/etc out of lightning module
- **Why**: To be able to train LVAE based algorithms in CAREamics
- **How**:  Coded with my elbows, unfortunately

### Changes Made

 **Added**:  
- Some amount of very questionable quality code which needs to be refactored right away
- Very basic tests which just ensure loss is not nan

### Known problems

- Noise models doesn't work because even the logic is actually not there yet
- Training logic for NMs isn't clear
- Most of the parameters should be documented, some should be refactored or removed 
- The whole pipeline isn't complete

---

**Please ensure your PR meets the following requirements:**

- [ ] Code builds and passes tests locally, including doctests
- [ ] New tests have been added (for bug fixes/features)
- [ ] Pre-commit passes
- [ ] PR to the documentation exists (for bug fixes / features)

### CODE REVIEW

File: src/careamics/careamist.py [hunk 1]
```diff
@@ -150,9 +151,24 @@ def __init__(
...
+                == SupportedArchitecture.UNET
+            ):
+                self.model: FCNModule = FCNModule(
+                    algorithm_config=self.cfg.algorithm_config,
+                )
+            elif (
+                self.cfg.algorithm_config.model.architecture
+                == SupportedArchitecture.LVAE
+            ):  # TODO fuckin mypy thinks model "already defined on line 158"
+                self.model: VAEModule = VAEModule(
```
Comment (jdeschamps) [inline src/careamics/careamist.py]: I believe what you can do about this is to define a type at the top of the module:
```
Module = Union[FCNModule, VAEModule]
```

And use that type on line 158.

File: src/careamics/careamist.py [hunk 2]
```diff
@@ -150,9 +151,24 @@ def __init__(
             self.cfg = source
 
             # instantiate model
-            self.model = CAREamicsModule(
-                algorithm_config=self.cfg.algorithm_config,
-            )
+            if (
```
Comment (jdeschamps) [inline src/careamics/careamist.py:153]: It would probably makes sense to refactor somewhere else the module creation since the same if-else statement repeats below!

File: src/careamics/careamist.py [hunk 3]
```diff
@@ -150,9 +151,24 @@ def __init__(
...
+                    algorithm_config=self.cfg.algorithm_config,
+                )
+            elif (
+                self.cfg.algorithm_config.model.architecture
+                == SupportedArchitecture.LVAE
+            ):  # TODO fuckin mypy thinks model "already defined on line 158"
+                self.model: VAEModule = VAEModule(
+                    algorithm_config=self.cfg.algorithm_config,
+                )
+            else:
```
Comment (jdeschamps) [inline src/careamics/careamist.py:160]: The configuration allows custom models that are neither Unet nor LVAE (`Custom` I believe). This breaks that possibility.

We could potentially consider that they should be `FCNModule` by default. Not sure what it entails.

File: src/careamics/careamist.py (hunk 2 above)
Comment (jdeschamps) [reply src/careamics/careamist.py:153]: But for now I think the `CAREamist` should just not be able to create a `LVAEModule` at all, since it is not possible to train it properly.

File: src/careamics/config/algorithm_model.py [hunk 4]
```diff
@@ -92,21 +94,23 @@ class AlgorithmConfig(BaseModel):
...
-    algorithm: Literal["n2v", "care", "n2n", "custom"]  # defined in SupportedAlgorithm
-    """Name of the algorithm, as defined in SupportedAlgorithm."""
-
-    loss: Literal["n2v", "mae", "mse"]
-    """Loss function to use, as defined in SupportedLoss."""
-
-    model: Union[UNetModel, VAEModel, CustomModel] = Field(discriminator="architecture")
-    """Model architecture to use, defined in SupportedArchitecture."""
+    # defined in SupportedAlgorithm
+    algorithm: Literal["n2v", "care", "n2n", "hdn", "musplit", "denoisplit", "custom"]
```
Comment (jdeschamps) [inline src/careamics/config/algorithm_model.py]: The overall configuration should not allow for the algorithms that don't work yet!

File: src/careamics/config/algorithm_model.py [hunk 5]
```diff
@@ -146,11 +150,54 @@ def algorithm_cross_validation(self: Self) -> Self:
...
+        noise_model : GaussianMixtureNoiseModel
+            Noise model.
+
+        Returns
+        -------
+        GaussianMixtureNoiseModel
+            Validated noise model.
+        """
+        if noise_model is not None:
+            try:
```
Comment (jdeschamps) [inline src/careamics/config/algorithm_model.py]: This is not clear to me, if `GaussianMixtureNoiseModel` is a Pydantic model, why wouldn't it have a `type` (if `type` is defined in the model that is)?

File: src/careamics/config/architectures/lvae_model.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+from typing_extensions import Self
+
+from .architecture_model import ArchitectureModel
+
+
+class LVAEModel(ArchitectureModel):
+    """LVAE model."""
+
+    model_config = ConfigDict(
+        use_enum_values=True, protected_namespaces=(), validate_assignment=True
```
Comment (jdeschamps) [inline src/careamics/config/architectures/lvae_model.py]: Some leftovers from old code probably:
- `use_enum_values` can go since we are not using `Enum` here
- `protected_namespaces` as well since no parameter is called `model_*`

File: src/careamics/config/architectures/lvae_model.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    )
+
+    architecture: Literal["LVAE"]
+    input_shape: int = Field(default=64, ge=8, le=1024)
```
Comment (jdeschamps) [inline src/careamics/config/architectures/lvae_model.py:17]: Can you add documentation for each parameter. A simple for now that we can review with Ashesh.

See the recent changes in the rest of the Pydantic models.

e.g.:
```   python
input_shape: int = Field(default=64, ge=8, le=1024)
"""Dimension of the input tensor along X and Y."""
```

File: src/careamics/config/architectures/lvae_model.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    multiscale_count: int = Field(default=3)  # TODO clarify
+    # 0 - off, len(z_dims) - 1
+    z_dims: tuple = Field(
+        default=(128, 128, 128, 128), validate_default=True
+    )  # TODO don't use tuple
```
Comment (jdeschamps) [inline src/careamics/config/architectures/lvae_model.py]: That's a TODO that should be easy to have in here and not leave hanging. Or is `mypy` complaining everywhere?

File: src/careamics/config/architectures/lvae_model.py (hunk 6 above)
Comment (jdeschamps) [reply src/careamics/config/architectures/lvae_model.py]: if all the `Field` have `validate_default=True`, then it can also be moved to the `ConfigDict` and removed from the `Field`.

File: src/careamics/config/architectures/lvae_model.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+    predict_logvar: Literal[None, "global", "pixelwise", "channelwise"] = (
+        None  # TODO Remove ?
+    )
+    enable_noise_model: bool = Field(default=True, validate_default=True)
+    analytical_kl: bool = Field(default=False, validate_default=True)
+
+    @field_validator("encoder_n_filters")
+    @classmethod
+    def validate_encoder_n_filters(cls, encoder_n_filters: int) -> int:
```
Comment (jdeschamps) [inline src/careamics/config/architectures/lvae_model.py]: Can you rename the method to `validate_even`? It makes more sense to me that the validator name is about the validation constraints rather than the variable it validate.

At some points (not now), we will have to go through all our validators and see whether we can refactor some outside. We may already be using a `validate_even` somewhere. And maybe check (not now), this type constraints libraries Talley presented.

File: src/careamics/config/architectures/lvae_model.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+            raise ValueError(
+                f"Number of channels for the bottom layer must be even"
+                f" (got {encoder_n_filters})."
+            )
+
+        return encoder_n_filters
+
+    @field_validator("decoder_n_filters")
+    @classmethod
+    def validate_decoder_n_filters(cls, decoder_n_filters: int) -> int:
```
Comment (jdeschamps) [inline src/careamics/config/architectures/lvae_model.py]: This can use the previous validator:

``` python
@field_validator("encoder_n_filters", "decoder_n_filters")
```

File: src/careamics/config/architectures/lvae_model.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        if decoder_n_filters % 2 != 0:
+            raise ValueError(
+                f"Number of channels for the bottom layer must be even"
+                f" (got {decoder_n_filters})."
+            )
+
+        return decoder_n_filters
+
+    @field_validator("z_dims")
+    def validate_z_dims(cls, z_dims: tuple) -> tuple:
```
Comment (jdeschamps) [inline src/careamics/config/architectures/lvae_model.py:103]: This can be done simply in the `Field` with the parameter `min_length` as far as I remember.

File: src/careamics/config/likelihood_model.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+class GaussianLikelihoodModel(BaseModel):
+    """Gaussion likelihood model.
+
+    Parameters
+    ----------
+    BaseModel
+    """
+
+    model_config = ConfigDict(validate_assignment=True, arbitrary_types_allowed=True)
```
Comment (jdeschamps) [inline src/careamics/config/likelihood_model.py]: I don't see any arbitrary type, so that parameters can be removed from the `ConfigDict`.

File: src/careamics/config/likelihood_model.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+    type: Literal["NMLikelihoodModel"]
+
+    ch_in: int
+    color_channels: int
+    data_mean: Union[dict[str, torch.Tensor], torch.Tensor]
```
Comment (jdeschamps) [inline src/careamics/config/likelihood_model.py:47]: This will not be serializable and deserializable in this state.

File: src/careamics/config/algorithm_model.py [hunk 14]
```diff
@@ -92,21 +94,23 @@ class AlgorithmConfig(BaseModel):
...
+    )
 
     # Optional fields
     optimizer: OptimizerModel = OptimizerModel()
     """Optimizer to use, defined in SupportedOptimizer."""
 
     lr_scheduler: LrSchedulerModel = LrSchedulerModel()
-    """Learning rate scheduler to use, defined in SupportedScheduler."""
+    # TODO spit algorithmConfig into separate classes because of this ?
+    likelihood: Optional[Union[GaussianLikelihoodModel, NMLikelihoodModel]] = Field(
```
Comment (jdeschamps) [inline src/careamics/config/algorithm_model.py]: If we are not using them, then we should not put them yet.

For instance I see that the `Likelihood` have tensors declared, that means we will probably not be able to write the configuration to disk or read it.

File: src/careamics/config/nm_model.py [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+from typing import Any, Literal
+
+from pydantic import BaseModel, ConfigDict, Field
+
+
+class GaussianMixtureNoiseModel(BaseModel):
+    """Gaussian mixture noise model."""
+
+    model_config = ConfigDict(validate_assignment=True, arbitrary_types_allowed=True)
```
Comment (jdeschamps) [inline src/careamics/config/nm_model.py]: Here probably `arbitrary_types_allowed` is not necessary either? Although I don't know how Pydantic deals with `Any`.

File: src/careamics/config/support/supported_algorithms.py [hunk 16]
```diff
@@ -15,6 +15,9 @@ class SupportedAlgorithm(str, BaseEnum):
     CARE = "care"
     N2N = "n2n"
     CUSTOM = "custom"
+    HDN = "hdn"
```
Comment (jdeschamps) [inline src/careamics/config/support/supported_algorithms.py]: Too early!

File: src/careamics/lightning/lightning_module.py [hunk 17]
```diff
@@ -202,6 +205,200 @@ def configure_optimizers(self) -> Any:
...
+        -------
+        Any
+            Loss value.
+        """
+        x, *aux = batch
+        out = self.model(x)
+        target = aux[0]
+        self.loss_parameters.current_epoch = self.current_epoch
+        self.loss_parameters.inputs = x
+        self.loss_parameters.mask = ~((target == 0).reshape(len(target), -1).all(dim=1))
```
Comment (jdeschamps) [inline src/careamics/lightning/lightning_module.py:303]: We would need some explanations here, because it is not clear what this is doing! Maybe add a TODO for extended comment....

File: src/careamics/lightning/lightning_module.py [hunk 18]
```diff
@@ -202,6 +205,200 @@ def configure_optimizers(self) -> Any:
...
+
+        Returns
+        -------
+        Any
+            Loss value.
+        """
+        x, *aux = batch
+        out = self.model(x)
+        target = aux[0]
+        self.loss_parameters.current_epoch = self.current_epoch
```
Comment (jdeschamps) [inline src/careamics/lightning/lightning_module.py:301]: Could the `loss_parameters` be an object and have a `self.loss_parameters.update(...)` to refactor these steps somewhere else?

If that makes sense, you could leave a TODO here.

File: src/careamics/losses/loss_factory.py [hunk 19]
```diff
@@ -4,20 +4,86 @@
...
 
-# TODO add tests
-# TODO add custom?
 def loss_factory(loss: Union[SupportedLoss, str]) -> Callable:
     """Return loss function.
 
     Parameters
     ----------
-    loss : Union[SupportedLoss, str]
+    loss : Union[SupportedLoss: str]
```
Comment (jdeschamps) [inline src/careamics/losses/loss_factory.py]: i don't understand the `:`?

File: src/careamics/losses/loss_factory.py [hunk 20]
```diff
@@ -4,20 +4,86 @@
...
+from torch import Tensor as tensor
+
 from ..config.support import SupportedLoss
-from .losses import mae_loss, mse_loss, n2v_loss
+from .fcn.losses import mae_loss, mse_loss, n2v_loss
+from .lvae.losses import denoisplit_loss, musplit_loss
+
+
+@dataclass
+class FCNLossParameters:
```
Comment (jdeschamps) [inline src/careamics/losses/loss_factory.py:18]: We would need description for each of these parameters.

File: src/careamics/losses/lvae/loss_utils.py [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+    Takes in the KL with shape (batch size, layers), returns the KL with
+    free bits (for optimization) with shape (layers,), which is the average
+    free-bits KL per layer in the current batch.
+    If batch_average is False (default), the free bits are per layer and
+    per batch element. Otherwise, the free bits are still per layer, but
+    are assigned on average to the whole batch. In both cases, the batch
+    average is returned, so it's simply a matter of doing mean(clamp(KL))
+    or clamp(mean(KL)).
+
+    Args:
```
Comment (jdeschamps) [inline src/careamics/losses/lvae/loss_utils.py:24]: This does not abide to the same docstring style as the rest of the code base!

File: src/careamics/losses/lvae/loss_utils.py [hunk 22]
```diff
@@ -0,0 +1,61 @@
+import torch
+
```
Comment (jdeschamps) [inline src/careamics/losses/lvae/loss_utils.py:2]: Add TODOs for tests

File: src/careamics/losses/lvae/losses.py [hunk 23]
```diff
@@ -0,0 +1,377 @@
+##### REQUIRED Methods for Loss Computation #####
```
Comment (jdeschamps) [inline src/careamics/losses/lvae/losses.py:1]: Module docstring?

File: src/careamics/losses/lvae/losses.py [hunk 24]
```diff
@@ -0,0 +1,377 @@
...
+from typing import TYPE_CHECKING
+
+import numpy as np
+import torch
+
+if TYPE_CHECKING:
+    print("sdfdsfds")
```
Comment (jdeschamps) [inline src/careamics/losses/lvae/losses.py]: That's an interesting if statement. 😆

File: src/careamics/losses/lvae/losses.py [hunk 25]
```diff
@@ -0,0 +1,377 @@
...
+    likelihood_obj: LikelihoodModule,
+    splitting_mask: torch.Tensor = None,
+    return_predicted_img: bool = False,
+) -> dict[str, torch.Tensor]:
+    """
+    Function to compute the reconstruction loss.
+
+    Parameters
+    ----------
+    reconstruction: torch.Tensor,
```
Comment (jdeschamps) [inline src/careamics/losses/lvae/losses.py:23]: Missing parameters descriptions and return section!

File: src/careamics/losses/lvae/losses.py [hunk 26]
```diff
@@ -0,0 +1,377 @@
...
+    target: torch.Tensor
+    input: torch.Tensor
+    splitting_mask: torch.Tensor = None
+        A boolean tensor that indicates which items to keep for reconstruction loss computation.
+        If `None`, all the elements of the items are considered (i.e., the mask is all `True`).
+    return_predicted_img: bool = False
+    likelihood_obj: LikelihoodModule = None
+    """
+    output = _get_reconstruction_loss_vector(
```
Comment (jdeschamps) [inline src/careamics/losses/lvae/losses.py:32]: Is all this worth commenting on or should it be subject to later refactoring?

File: src/careamics/models/lvae/likelihoods.py [hunk 27]
```diff
@@ -9,6 +9,42 @@
...
+    ----------
+    config: Union[GaussianLikelihoodModel, NMLikelihoodModel]
+        The configuration object for the likelihood module.
+
+    Returns
+    -------
+    nn.Module
+        The likelihood module.
+    """
+    if config.type == "GaussianLikelihoodModel":
```
Comment (jdeschamps) [inline src/careamics/models/lvae/likelihoods.py]: it looks much better like this 😆

File: src/careamics/models/lvae/lvae.py [hunk 28]
```diff
@@ -4,74 +4,71 @@
 The current implementation is based on "Interpretable Unsupervised Diversity Denoising and Artefact Removal, Prakash et al."
 """
 
-from typing import Dict, Iterable, List, Tuple, Union
+from collections.abc import Iterable
+from typing import Dict, List, Tuple
 
-import ml_collections
+import ml_collections  # TODO: refactor this out
```
Comment (jdeschamps) [inline src/careamics/models/lvae/lvae.py]: I don't see it used except in the declaration of `config: ml_collections` and then this `config` is not used (as far as I can tell). So kick it out as we discussed to allow the tests from passing (hopefully)!

File: src/careamics/models/lvae/noise_models.py [hunk 29]
```diff
@@ -1,12 +1,45 @@
...
+    """
+    if model_config:
+        if model_config.type == "GaussianMixtureNoiseModel":
+            return GaussianMixtureNoiseModel(**model_config.model_dump())
+        else:
+            raise NotImplementedError(f"Model {model_config.type} is not implemented")
+    return None
+
 
 class DisentNoiseModel(nn.Module):
```
Comment (jdeschamps) [inline src/careamics/models/lvae/noise_models.py:46]: Do we know what this is?

**Review by jdeschamps — CHANGES_REQUESTED** [review]
I think we should stick with the original plan of only using the new algorithms (HDN, denoiSplit) with the `LightningAPI` and not make them available through the `CAREamist AP`.

We are very close to a `0.1.0` release, which means that we might get an influx of users and incorporating code that does not work or might break can create a lot of issues.

If the modules are in the code base, they can be instantiated independently with the `Ligthning API`, that will be more than enough for Ashesh's paper. Once we are confident the quality of the code (integration, testing, documentation) is high enough, then we can make it available through the `CAREamist`.

So in the meantime, I would suggest keeping the configuration clean (removing LVAE, NoiseModels, Likelihood, and the new entries in the supported enums) and adding placeholder `NotImplementedErrors` in `CAREamist`.

For now, `NoiseModel` and `Likelihood` could be passed directly as parameters along the algorithm configuration. We can also discuss alternatives (e.g. allowing additional parameters in the Pydantic model via `ConfigDict`, but that would require a lot of `asserts` in the LVAE related code).

File: src/careamics/careamist.py (hunk 1 above)
Comment (CatEek) [reply src/careamics/careamist.py]: But mypy isn't complaining about type ?

File: src/careamics/careamist.py (hunk 2 above)
Comment (CatEek) [reply src/careamics/careamist.py:153]: Absolutely. I put `notimplemented` there for now

File: src/careamics/careamist.py (hunk 3 above)
Comment (CatEek) [reply src/careamics/careamist.py:160]: Let's discuss

File: src/careamics/config/algorithm_model.py (hunk 4 above)
Comment (CatEek) [reply src/careamics/config/algorithm_model.py]: But then how would you expect tests to pass ?

File: src/careamics/config/algorithm_model.py (hunk 5 above)
Comment (CatEek) [reply src/careamics/config/algorithm_model.py]: Not clear to me neither, NMs still need lot's of work

File: src/careamics/config/architectures/lvae_model.py (hunk 7 above)
Comment (CatEek) [reply src/careamics/config/architectures/lvae_model.py:17]: That's what I asked @ashesh-0 and @federico-carrara to do

File: src/careamics/config/architectures/lvae_model.py (hunk 11 above)
Comment (CatEek) [reply src/careamics/config/architectures/lvae_model.py:103]: There will be some other constrains, that's why i left it there for now

File: src/careamics/config/likelihood_model.py (hunk 13 above)
Comment (CatEek) [reply src/careamics/config/likelihood_model.py:47]: Don't know yet

File: src/careamics/losses/loss_factory.py (hunk 20 above)
Comment (CatEek) [reply src/careamics/losses/loss_factory.py:18]: Same as with other descriptions

File: src/careamics/losses/lvae/loss_utils.py (hunk 21 above)
Comment (CatEek) [reply src/careamics/losses/lvae/loss_utils.py:24]: Same

File: src/careamics/losses/lvae/losses.py (hunk 24 above)
Comment (CatEek) [reply src/careamics/losses/lvae/losses.py]: ahahahah

File: src/careamics/models/lvae/noise_models.py (hunk 29 above)
Comment (CatEek) [reply src/careamics/models/lvae/noise_models.py:46]: No

**Comment by melisande-c** [comment]
> So in the meantime, I would suggest keeping the configuration clean (removing LVAE, NoiseModels, Likelihood, and the new entries in the supported enums) and adding placeholder `NotImplementedErrors` in `CAREamist`.

I will agree and say let's make the minimum number of modifications possible to the rest of the code until the LVAE pipeline is more complete, maybe not even adding not-implemented-errors to `CAREamist`. It might become clearer how we add the new components into the rest of `CAREamist` and the configuration once we understand how they work. I think it will also make the changes to the code easier to follow.

File: src/careamics/careamist.py (hunk 1 above)
Comment (jdeschamps) [reply src/careamics/careamist.py]: One needs to try, I haven't. 😄

File: src/careamics/config/algorithm_model.py (hunk 4 above)
Comment (jdeschamps) [reply src/careamics/config/algorithm_model.py]: The tests should then not be implemented to test those until they are officially supported!

**Comment by jdeschamps** [comment]
It seems there is still a `hdn` hanging around in one of the tests!

File: src/careamics/careamist.py [hunk 30]
```diff
@@ -166,9 +184,15 @@ def __init__(
...
-                )
+                if (
+                    self.cfg.algorithm_config.model.architecture
+                    == SupportedArchitecture.UNET
+                ):
+                    self.model = FCNModule(
+                        algorithm_config=self.cfg.algorithm_config,
+                    )
+                else:
+                    raise NotImplementedError("Architecture not supported.")
```
Comment (melisande-c) [inline src/careamics/careamist.py:185]: Repetition of inner if-else statement can be removed through separation of config initialisation and model initialisation. i.e.

```python
if isinstance(source, Path) and source.is_file():
    if source.suffix == ".yaml" or source.suffix == ".yml":
        self.cfg = load_configuration(source)
        self.model = None
    else:
        self.model, self.cfg = load_pretrained(source)
elif isinstance(source, Configuration):
    self.cfg = source
    self.model = None
else:
    raise ValueError("appropriate message")

# instantiate model
if self.model is None:
    if (
        self.cfg.algorithm_config.model.architecture
        == SupportedArchitecture.UNET
    ):
        self.model = FCNModule(
            algorithm_config=self.cfg.algorithm_config,
        )
    elif (
        self.cfg.algorithm_config.model.architecture
        == SupportedArchitecture.LVAE
    ):  
        raise NotImplementedError(
            "LVAE is currently only supported through lightning api"
        )
    else:
        raise NotImplementedError("Architecture not supported.")
```

^ just an example can probably be improved further. Extracting to smaller testable sub functions might be good too.

File: src/careamics/config/algorithm_model.py [hunk 31]
```diff
@@ -89,24 +89,20 @@ class AlgorithmConfig(BaseModel):
...
-    model: Union[UNetModel, VAEModel, CustomModel] = Field(discriminator="architecture")
-    """Model architecture to use, defined in SupportedArchitecture."""
+    model: Union[UNetModel, CustomModel] = Field(discriminator="architecture")
 
     # Optional fields
     optimizer: OptimizerModel = OptimizerModel()
     """Optimizer to use, defined in SupportedOptimizer."""
 
     lr_scheduler: LrSchedulerModel = LrSchedulerModel()
-    """Learning rate scheduler to use, defined in SupportedScheduler."""
```
Comment (melisande-c) [inline src/careamics/config/algorithm_model.py]: Some docstrings have been deleted, probably by mistake.

File: src/careamics/lightning/lightning_module.py [hunk 32]
```diff
@@ -202,6 +205,201 @@ def configure_optimizers(self) -> Any:
         }
 
 
+class VAEModule(L.LightningModule):
```
Comment (melisande-c) [inline src/careamics/lightning/lightning_module.py:208]: Are we assuming LVAE and VAE will be able to share a lightning module? (Just a question for my understanding)

File: src/careamics/config/likelihood_model.py [hunk 33]
```diff
@@ -0,0 +1,45 @@
...
+    data_std: Union[dict[str, torch.Tensor], torch.Tensor]
+    noiseModel: nn.Module
```
Comment (melisande-c) [inline src/careamics/config/likelihood_model.py:49]: Camel casing of `noiseModel` doesn't adhere to python style.

File: src/careamics/config/support/supported_architectures.py [hunk 34]
```diff
@@ -14,7 +14,7 @@ class SupportedArchitecture(str, BaseEnum):
     """
 
     UNET = "UNet"
-    VAE = "VAE"
+    LVAE = "LVAE"
     CUSTOM = (
-        "Custom"  # TODO all the others tags are small letters, except the architect
+        "custom"  # TODO all the others tags are small letters, except the architect
```
Comment (melisande-c) [inline src/careamics/config/support/supported_architectures.py]: can remove TODO

File: src/careamics/lightning/lightning_module.py [hunk 35]
```diff
@@ -202,6 +205,201 @@ def configure_optimizers(self) -> Any:
...
+        self.loss_parameters.inputs = x
+        self.loss_parameters.mask = ~((target == 0).reshape(len(target), -1).all(dim=1))
+        self.loss_parameters.likelihood = self.likelihood
+        self.loss_parameters.noise_model = self.noise_model
+        loss = self.loss_func(out, target, self.loss_parameters)  # TODO ugly ?
+
+        self.log_dict(loss, on_step=True, on_epoch=True, prog_bar=True, logger=True)
+        return loss
+
+    def validation_step(self, batch: Tensor, batch_idx: Any) -> None:
```
Comment (melisande-c) [inline src/careamics/lightning/lightning_module.py:311]: Seems like a lot has been removed from the validation step in src/careamics/lvae_training/lightning_module.py is everything in there definitely useless?

File: src/careamics/models/lvae/layers.py [hunk 36]
```diff
@@ -117,7 +117,7 @@ def __init__(
                     bias=conv2d_bias,
                 )
                 modules.append(conv)
-                modules.append(nonlin())
+                modules.append(nonlin)
```
Comment (melisande-c) [inline src/careamics/models/lvae/layers.py:120]: if `nonlin=nn.ReLU` as docstring suggests then this should be `module.append(nonlin())`

File: src/careamics/config/architectures/lvae_model.py [hunk 37]
```diff
@@ -0,0 +1,173 @@
...
+    decoder_dropout: float = Field(default=0.1, ge=0.0, le=0.9)
+    nonlinearity: Literal[
+        "None", "Sigmoid", "Softmax", "Tanh", "ReLU", "LeakyReLU", "ELU"
+    ] = Field(
+        default="ELU",
+    )
+
+    predict_logvar: Literal[None, "global", "pixelwise", "channelwise"] = (
+        None  # TODO Remove ?
+    )
```
Comment (melisande-c) [inline src/careamics/config/architectures/lvae_model.py:34]: If creation of likelihood models has been extracted to `likelihood_factory` then yes it seems this parameter can be removed from LVAE.

File: src/careamics/models/lvae/layers.py (hunk 36 above)
Comment (melisande-c) [reply src/careamics/models/lvae/layers.py:120]: Ok nvm, I see the `get_activation` function returns for example `nn.ReLU()`. In which case the docstring should be updated.

File: src/careamics/models/lvae/lvae.py [hunk 38]
```diff
@@ -494,29 +447,28 @@ def create_likelihood_module(self):
...
             )
-        if self.loss_type == LossType.DenoiSplitMuSplit or self.likelihood_NM is None:
+        else:
+            self.likelihood_gm = GaussianLikelihood(
+                self.decoder_n_filters,
+                self.target_ch,
+                predict_logvar=self.predict_logvar,
+                logvar_lowerbound=self.logvar_lowerbound,
+                conv2d_bias=self.topdown_conv2d_bias,
+            )
```
Comment (melisande-c) [inline src/careamics/models/lvae/lvae.py:466]: When is `self.likelihood_gm` used? as far as I can see it doesn't seem to be used by any of the`LVAE` class' methods. If it is only used by external functions maybe it can be removed from the `LVAE` class and passed to those functions explicitly.

File: tests/config/architectures/test_lvae_model.py [hunk 39]
```diff
@@ -0,0 +1,134 @@
...
+
+
+def test_instantiation():
+    """Test that LVAEModel can be instantiated."""
+    model_params = {
+        "architecture": "LVAE",
+    }
+
+    # instantiate model
+    LVAEModel(**model_params)
```
Comment (melisande-c) [inline tests/config/architectures/test_lvae_model.py:14]: maybe assert attributes are set as expected.

File: tests/config/architectures/test_lvae_model.py [hunk 40]
```diff
@@ -0,0 +1,134 @@
...
+
+    # Algorithm json schema
+    schema = LVAEModel.model_json_schema()
+
+    # check that all activations are supported
+    for act in schema["properties"]["nonlinearity"]["enum"]:
+        assert act in activations
+
+
+def test_activation_wrong_values():
```
Comment (melisande-c) [inline tests/config/architectures/test_lvae_model.py:84]: maybe rename `test_activation_unknown_value`

File: tests/config/test_algorithm_model.py [hunk 41]
```diff
@@ -56,7 +49,7 @@ def test_model_discriminator(minimum_algorithm_n2v):
     "algorithm, loss, model",
     [
         ("n2v", "n2v", {"architecture": "UNet", "n2v2": False}),
-        ("custom", "mae", {"architecture": "UNet", "n2v2": True}),
+        ("n2n", "mae", {"architecture": "UNet", "n2v2": False}),
```
Comment (melisande-c) [inline tests/config/test_algorithm_model.py:68]: why this change in parametrisation?

File: tests/conftest.py [hunk 42]
```diff
@@ -22,7 +22,7 @@ def create_tiff(path: Path, n_files: int):
 
 
 @pytest.fixture
-def minimum_algorithm_custom() -> dict:
+def minimum_algorithm_n2v() -> dict:
```
Comment (melisande-c) [inline tests/conftest.py:70]: The git comparison has got a bit messed up, but has custom algorithm been removed?

**Review by melisande-c — COMMENTED** [review]
So I still think there are some changes here that are out of scope for the PR (also the current PR title is not descriptive of the changes). I see though that it's pretty hard to make the pydantic model when the `LVAE` class does need some refactoring to extract components that don't necessarily need to be contained within the class. Maybe the splitting of the configuration you mentioned will help.

**Comment by jdeschamps** [comment]
I pushed a fix for the doctests as promised.

File: tests/conftest.py (hunk 42 above)
Comment (jdeschamps) [reply tests/conftest.py:70]: Looking into the code, it seems that the rational for `custom` algorithm was to make sure to escape any assumption on the model parameters: https://github.com/CAREamics/careamics/blob/aed4d82387590159a005f3e8d6e08fc8efad168f/src/careamics/config/configuration_model.py#L219

That means that it should not be removed!

Comment (jdeschamps) [reply tests/conftest.py:70]: I added back the `custom` algorithm, and added tests that were missing to test what it was actually used for.

**Comment by jdeschamps** [comment]
Comment https://github.com/CAREamics/careamics/pull/180/files#r1675427899 still needs to be fixed regarding the `custom` algorithms and models.

File: src/careamics/careamist.py (hunk 3 above)
Comment (jdeschamps) [reply src/careamics/careamist.py:160]: Now that the configuration has only `custom` and `UNet` models, these statements should go.

Once the `LVAE` is in the configuration, we will need to reassess how to deal with `custom` models.

File: src/careamics/lightning/lightning_module.py (hunk 32 above)
Comment (jdeschamps) [reply src/careamics/lightning/lightning_module.py:208]: I guess so, VAE will just be a LVAE with one level.

File: src/careamics/lightning/lightning_module.py (hunk 17 above)
Comment (jdeschamps) [reply src/careamics/lightning/lightning_module.py:303]: Please add comments about what is happening in these steps.

**Review by jdeschamps — CHANGES_REQUESTED** [review]

File: src/careamics/config/algorithm_model.py [hunk 43]
```diff
@@ -89,24 +89,29 @@ class AlgorithmConfig(BaseModel):
...
-    algorithm: Literal["n2v", "care", "n2n", "custom"]  # defined in SupportedAlgorithm
-    """Name of the algorithm, as defined in SupportedAlgorithm."""
+    # defined in SupportedAlgorithm
+    algorithm: Literal["n2v", "care", "n2n", "custom"]
+    """Name of the algorithm, as defined in SupportedAlgorithm. Use `custom` for custom
+    model architecture."""
 
     loss: Literal["n2v", "mae", "mse"]
-    """Loss function to use, as defined in SupportedLoss."""
+    """Name of the algorithm, as defined in SupportedAlgorithm."""
```
Comment (melisande-c) [inline src/careamics/config/algorithm_model.py:102]: Algorithm docs instead of loss docs.

File: src/careamics/config/algorithm_model.py [hunk 44]
```diff
@@ -146,8 +151,10 @@ def algorithm_cross_validation(self: Self) -> Self:
             if self.loss == "n2v":
                 raise ValueError("Supervised algorithms do not support loss `n2v`.")
 
-        if isinstance(self.model, VAEModel):
-            raise ValueError("VAE are currently not implemented.")
+        if (self.algorithm == "custom") != (self.model.architecture == "custom"):
+            raise ValueError(
+                "Algorithm and model architecture must be both `custom` or not."
```
Comment (melisande-c) [inline src/careamics/config/algorithm_model.py:156]: grammar could be clearer. not what?

File: src/careamics/careamist.py (hunk 30 above)
Comment (CatEek) [reply src/careamics/careamist.py:185]: Those things are gone in the config refac pr

File: src/careamics/lightning/lightning_module.py (hunk 17 above)
Comment (CatEek) [reply src/careamics/lightning/lightning_module.py:303]: I wish I knew, it's taken from the original code as is

File: src/careamics/lightning/lightning_module.py (hunk 35 above)
Comment (CatEek) [reply src/careamics/lightning/lightning_module.py:311]: It wasn't added, not "was removed"

File: tests/config/architectures/test_lvae_model.py (hunk 40 above)
Comment (CatEek) [reply tests/config/architectures/test_lvae_model.py:84]: I don't know what that function even doing

File: src/careamics/lightning/lightning_module.py (hunk 35 above)
Comment (melisande-c) [reply src/careamics/lightning/lightning_module.py:311]: What I mean is in the lightning module `LadderVAELight` in this [file](https://github.com/CAREamics/careamics/blob/aed4d82387590159a005f3e8d6e08fc8efad168f/src/careamics/lvae_training/lightning_module.py), there is a lot more going on in the validation step.

File: tests/config/architectures/test_lvae_model.py (hunk 40 above)
Comment (melisande-c) [reply tests/config/architectures/test_lvae_model.py:84]: Just testing an error is raised if "nonlinearity" is not a supported activation, no?

File: src/careamics/config/algorithm_model.py (hunk 44 above)
Comment (jdeschamps) [reply src/careamics/config/algorithm_model.py:156]: I guess it was meant to be "None"!
